Convert delete room admin API to async endpoint (#11223)
matrix-org/synapse: docs/admin_api/rooms.md

to the new room will have power level `-10` by default, and thus be unable to speak.

If `block` is `true`, users will be prevented from joining the old room.

> This option can in [Version 1](#version-1-old-version) also be used to pre-emptively

Reviewer: wait, what? why are we removing this ability? It's important to be able to pre-emptively block rooms.

Author: That ist not removed. The ability in Version 1 is created in #11228 and there is also an issue #11283.
IMO a simple block of a room with a bg task is not a very good usage. And it is implemented with raise and catching exceptions. I think v1 of this API should be removed after publishing the block API.

Reviewer: hrm. Ok, let's leave it for now.

The JSON body must not be empty. The body must be at least `{}`.

## Status of deleting rooms

It is possible to query the status of the background task for deleting rooms.
The status can be queried up to 24 hours after completion of the task
or a restart of Synapse.

The API is:

```
GET /_synapse/admin/v2/rooms/<room_id>/delete_status
```

A response body like the following is returned:

```json
{
    "status": "active",
    "result": {
        "kicked_users": [
            "@foobar:example.com"
        ],
        "failed_to_kick_users": [],
        "local_aliases": [
            "#badroom:example.com",
            "#evilsaloon:example.com"
        ],
        "new_room_id": "!newroomid:example.com"
    }
}
```

**Parameters**

The following parameters should be set in the URL:

* `room_id` - The ID of the room.

**Response**

The following fields are returned in the JSON response body:

* `status` - The status will be one of:
  - `remove members` - The process is removing users from the `room_id`.
  - `active` - The process is purging the room from databse.
  - `complete` - The process has completed successfully.

>   - `failed` - The process is aborted, an error has occurred.

Author: I am not sure what do you mean?
Should there some more information? Or do not return
```jsonc
"result": {
    "kicked_users": [],
    "failed_to_kick_users": [],
    "local_aliases": [],
    "new_room_id": null
}
```

>   - `failed` - The process is aborted, an error has occurred.

Author: Is this related to?
- #6934

>   - `failed` - The process is aborted, an error has occurred.

Reviewer: Yes, I meant "we should return more information", and it seems to be the same as #6934. We can leave it for now if you like.

>   - `failed` - The process is aborted, an error has occurred.

Author: I have added an error message of it fails.

The JSON body must not be empty. The body must be at least `{}`.

## Status of deleting rooms

**Note**: This API is new, experimental and "subject to change".

It is possible to query the status of the background task for deleting rooms.
The status can be queried up to 24 hours after completion of the task,
or until Synapse is restarted (whichever happens first).

With this API you can get the status of all tasks the last 24h for this `room_id`.
If you want only get the status of one specific task, you can also use

> the [Purge status query API](purge_history_api.md#Purge-status-query).

Reviewer: It feels confusing to tell users to use the *purge* status query for the results of a *delete* operation. I think if this is a thing we need, we should add a `GET /_synapse/admin/v2/rooms/<room_id>/delete_status/<purge_id>` for it.

Author: What is with this: `GET /_synapse/admin/v2/rooms/delete_status/<purge_id>`?
Than you can request the status with `purge_id` and do not need the `room_id`.

Reviewer: > What is with this: `GET /_synapse/admin/v2/rooms/delete_status/<purge_id>`?

sounds fine.

The API is:

```
GET /_synapse/admin/v2/rooms/<room_id>/delete_status
```

A response body like the following is returned:

```json
{
    "delete_status": [
        {
            "purge_id": "purgeid1",
            "status": "failed",

>             "result": {

Reviewer: given this is just the result of the shutdown operation, maybe:

```suggestion
            "shutdown_result": {
```

                "kicked_users": [],
                "failed_to_kick_users": [],
                "local_aliases": [],
                "new_room_id": null
            }
        }, {
            "purge_id": "purgeid2",
            "status": "active",
            "result": {
                "kicked_users": [
                    "@foobar:example.com"
                ],
                "failed_to_kick_users": [],
                "local_aliases": [
                    "#badroom:example.com",
                    "#evilsaloon:example.com"
                ],
                "new_room_id": "!newroomid:example.com"
            }
        }
    ]
}
```

**Parameters**

The following parameters should be set in the URL:

* `room_id` - The ID of the room.

**Response**

The following fields are returned in the JSON response body:

- `delete_status` - An array of objects, each containing information about one task.
  Task objects contain the following fields:
  - `status` - The status will be one of:

>     - `remove members` - The process is removing users from the `room_id`.
>     - `active` - The process is purging the room from database.

Reviewer: `active` is confusing here. How about:

```suggestion
    - `shutting_down` - The process is removing users from the room.
    - `purging` - The process is purging the room and event data from database.
```

Author: If I change `active` to `purging` I change the `/purge_history_status` API.
If this is okay. I do this. Otherwise I have to create a new status.

Author: Sytest fails.
https://github.com/matrix-org/sytest/blob/57c856c0c03f41ef9a6d06c89ca898e594505a4b/tests/48admin.pl#L28

Should I create an patch for sytest?
`   }, while => sub { $_[0]->get eq 'active' || $_[0]->get eq 'purging' });`

Author: And we certainly need an upgrade notice.

Reviewer: Sorry, I think this needs to be specific to delete requests.

I think this, and the problem with `asdict`, are really symptoms of re-using `PurgeStatus` for a different type of request.

Author: Ok. I will create a new class `DeleteStatus` to solve the problem.

Reviewer: Thanks. Sorry this is turning into a bit of an epic!

The JSON body must not be empty. The body must be at least `{}`.

## Status of deleting rooms

**Note**: This API is new, experimental and "subject to change".

It is possible to query the status of the background task for deleting rooms.
The status can be queried up to 24 hours after completion of the task,
or until Synapse is restarted (whichever happens first).

### Query by `room_id`

> With this API you can get the status of all tasks the last 24h for this `room_id`.

Reviewer: ```suggestion
With this API you can get the status of all active deletion tasks, and all those completed in the last 24h,
for the given `room_id`.
```
